PWGHF/D2H/Tasks/taskD0.cxx

```diff
@@ -517,6 +564,117 @@ struct HfTaskD0 {
+      if (rejectionMask != 0) {
+        continue;
+      }
+      auto bc = collision.template bc_as<BCsType>();
+      upchelpers::FITInfo fitInfo{};
+      udhelpers::getFITinfo(fitInfo, bc, bcs, ft0s, fv0as, fdds);
+
+      GapType gap = GapType::DoubleGap;
+      if (bc.has_zdc()) {
+        auto zdc = bc.zdc();
+        qaRegistry.fill(HIST("Data/fitInfo/ampFT0A_vs_ampFT0C"), fitInfo.ampFT0A, fitInfo.ampFT0C);
+        qaRegistry.fill(HIST("Data/zdc/energyZNA_vs_energyZNC"), zdc.energyCommonZNA(), zdc.energyCommonZNC());
+        gap = hf_upc::determineGapType(fitInfo.ampFT0A, fitInfo.ampFT0C, zdc.energyCommonZNA(), zdc.energyCommonZNC(),
+                                       upcFT0AThreshold, upcFT0CThreshold, upcZDCThreshold);
+        qaRegistry.fill(HIST("Data/hUpcGapAfterSelection"), hf_upc::gapTypeToInt(gap));
+      }
+      if (hf_upc::isSingleSidedGap(gap)) {
+        int const gapTypeInt = hf_upc::gapTypeToInt(gap);
+        const auto thisCollId = collision.globalIndex();
+        const auto& groupedD0Candidates = candidates.sliceBy(candD0PerCollision, thisCollId);
+        float occ{-1.f};
+
+        for (const auto& candidate : groupedD0Candidates) {
+          if (yCandRecoMax >= 0. && std::abs(HfHelper::yD0(candidate)) > yCandRecoMax) {
+            continue;
+          }
+
+          float massD0 = HfHelper::invMassD0ToPiK(candidate);
+          float massD0bar = HfHelper::invMassD0barToKPi(candidate);
+          auto ptCandidate = candidate.pt();
+
+          if (candidate.isSelD0() >= selectionFlagD0) {
+            registry.fill(HIST("hMass"), massD0, ptCandidate);
+            registry.fill(HIST("hMassFinerBinning"), massD0, ptCandidate);
+            registry.fill(HIST("hMassVsPhi"), massD0, ptCandidate, candidate.phi());
+          }
+          if (candidate.isSelD0bar() >= selectionFlagD0bar) {
+            registry.fill(HIST("hMass"), massD0bar, ptCandidate);
+            registry.fill(HIST("hMassFinerBinning"), massD0bar, ptCandidate);
+            registry.fill(HIST("hMassVsPhi"), massD0bar, ptCandidate, candidate.phi());
+          }
+
+          // Fill THnSparse with structure matching taskDplus: [mass, pt, mlScores, cent, occ, gapType, FT0A, FT0C]
+          if constexpr (fillMl) {
+            auto fillTHnData = [&](float mass, int d0Type) {
+              std::vector<double> valuesToFill{static_cast<double>(mass), static_cast<double>(ptCandidate),
+                                               candidate.mlProbD0()[0], candidate.mlProbD0()[1], candidate.mlProbD0()[2],
+                                               static_cast<double>(HfHelper::yD0(candidate)), static_cast<double>(d0Type)};
+              if (storeCentrality) {
+                valuesToFill.push_back(centrality);
+              }
+              if (storeOccupancyAndIR) {
+                valuesToFill.push_back(occ);
+              }
```

Comment:
This changes the position of the elements pushed afterwards. Is it not a problem?